module/extension/Director/src/Director.cpp

```diff
@@ -0,0 +1,254 @@
+        else {
+            throw std::runtime_error("Causing statements must come after a Provides, Entering or Leaving statement");
+        }
+    }
+
+    Director::Director(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+
+        on<Configuration>("Director.yaml").then("Configure", [this](const Configuration& config) {
+            log_level = config["log_level"].as<NUClear::LogLevel>();
+        });
+
+        // Removes all the Providers for a reaction when it is unbound
+        on<Trigger<Unbind>, Sync<Director>>().then("Remove Provider", [this](const Unbind& unbind) {  //
+            remove_provider(unbind.id);
+        });
+
+        // Add a Provider
+        on<Trigger<ProvidesReaction>, Sync<Director>>().then("Add Provider", [this](const ProvidesReaction& provides) {
+            add_provider(provides);
+        });
+
+        // Add a when expression to this Provider
+        on<Trigger<WhenExpression>, Sync<Director>>().then("Add When", [this](const WhenExpression& when) {  //
+            add_when(when);
+        });
+
+        // Add a causing condition to this Provider
+        on<Trigger<CausingExpression>, Sync<Director>>().then("Add Causing", [this](const CausingExpression& causing) {
+            add_causing(causing);
+        });
+
+        // A task has arrived, either it's a root task and send it off, or build up our pack for when it's done
+        on<Trigger<DirectorTask>, Sync<DirectorTask>>().then(
+            "Director Task",
+            [this](std::shared_ptr<const DirectorTask> task) {
+                // Root level task, make the TaskPack immediately and send it off to be executed as a root task
+                if (providers.count(task->requester_id) == 0) {
+                    emit(std::make_unique<TaskPack>(TaskPack({task})));
+                }
+                // Check if this Provider is active and allowed to make subtasks
+                else if (providers[task->requester_id]->active) {
+                    pack_builder.emplace(task->requester_task_id, task);
+                }
+                else {
+                    // Throw an error so the user can see what a fool they are being
+                    log<NUClear::WARN>("The task",
+                                       task->name,
+                                       "cannot be executed as the Provider",
+                                       providers[task->requester_id]->reaction->identifier[0],
+                                       "is not active and cannot make subtasks");
+                }
+            });
+
+        // This reaction runs when a Provider finishes to send off the task pack to the main director
+        on<Trigger<ProviderDone>, Sync<DirectorTask>>().then("Package Tasks", [this](const ProviderDone& done) {
+            // Get all the tasks that were emitted by this provider and send it as a task pack
+            auto e = pack_builder.equal_range(done.requester_task_id);
```

Comment:
Somewhat irrelevant since it's returning an std::pair with two iterators in it.
Also it'll be an rvalue not an lvalue that is returned so making it a const reference will work because of a technicality in C++, but it's not a reference it'll actually be an rvalue in disguise

Reply:
ahh true. That makes sense